Src/PerformanceCollector/Shared/Implementation/QuickPulse/Process/PerfLib/CategorySample.cs

```diff
@@ -0,0 +1,258 @@
+                    }
+                }
+
+                if (sample == null)
+                {
+                    throw new InvalidOperationException("Could not find the counter " + counterNameIndex);
+                }
+
+                // now set up the InstanceNameTable.  
+                if (!isMultiInstance)
+                {
+                    throw new InvalidOperationException("Single instance categories are not supported");
+                }
+
+                string[] parentInstanceNames = null;
+                this.InstanceNameTable = new Dictionary<string, int>(instanceNumber, StringComparer.OrdinalIgnoreCase);
+                for (int i = 0; i < instanceNumber; i++)
+                {
+                    NativeMethods.PERF_INSTANCE_DEFINITION perfInstance = new NativeMethods.PERF_INSTANCE_DEFINITION();
+                    Marshal.PtrToStructure(dataRef, perfInstance);
+
+                    if (perfInstance.ParentObjectTitleIndex > 0 && parentInstanceNames == null)
+                    {
+                        parentInstanceNames = this.GetInstanceNamesFromIndex(perfInstance.ParentObjectTitleIndex);
+                    }
+
+                    string instanceName;
+                    if (parentInstanceNames != null && perfInstance.ParentObjectInstance >= 0
+                        && perfInstance.ParentObjectInstance < parentInstanceNames.Length - 1)
+                    {
+                        instanceName = parentInstanceNames[perfInstance.ParentObjectInstance] + "/"
+                                       + Marshal.PtrToStringUni((IntPtr)((long)dataRef + perfInstance.NameOffset));
+                    }
+                    else
+                    {
+                        instanceName = Marshal.PtrToStringUni((IntPtr)((long)dataRef + perfInstance.NameOffset));
+                    }
+
+                    // in some cases instance names are not unique (Process), same as perfmon, so generate a unique name
+                    string newInstanceName = instanceName;
+                    int newInstanceNumber = 1;
+                    while (true)
+                    {
+                        if (!this.InstanceNameTable.ContainsKey(newInstanceName))
+                        {
+                            this.InstanceNameTable[newInstanceName] = i;
+                            break;
+                        }
+                        else
+                        {
+                            newInstanceName = instanceName + "#" + newInstanceNumber.ToString(CultureInfo.InvariantCulture);
+                            ++newInstanceNumber;
+                        }
+                    }
+
+                    dataRef = (IntPtr)((long)dataRef + perfInstance.ByteLength);
+
+                    // we only need one counter
+                    ////foreach (CounterDefinitionSample s in samples)
+                    ////{
+                    ////    s.SetInstanceValue(i, dataRef);
+                    ////}
```

Comment:
I'm leaving this in for reference in case we need to extract more counters in the future. This is an important extension point, so I think this is useful for someone reading it.